### DISCUSSION TITLE
Configure Renovate Bot To Only Use Private Registry URL

### DISCUSSION THREAD

**Discussion by jsrice7391** [renovatebot/renovate#8033] [Suggest an Idea]
I am migrating my company's private package registry to Code Artifact. Our private Registry is setup such that the installer first looks for the package within the private registry. If it does not find it, it goes to the public registries.

I am having a hard time with the `registryUrls` parameter within the configuration file. Right now I have this: 
```json
 "npm": {
    "registryUrls": [
      "//private_npm_registyr_url_with_token",
    ],
  }
```
When I run `npm install` from the project folder, it installs the package without issue. When I go to run a renovate scan, I get a 404 saying that a private package was not found:
```json
{
"message": "Response code 404 (Not Found)",
           "stack": "HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/Users/justinrice/Documents/IndigoGit/renovate-service/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:93:5)",
           "options": {
             "headers": {
               "user-agent": "https://github.com/renovatebot/renovate",
               "accept-encoding": "gzip, deflate",
               "cache-control": "no-cache",
               "accept": "application/json"
             },
             "url": "https://registry.npmjs.org/@indigo-ag%2Fprivate_sdk_name"
}
```
This is the expected behavior because the package is not published globally on npm, it's within our private Code Artifact registry.

The question becomes:
**- Is there a way to get renovate-bot to ONLY ever look at the registryUrl and not look for any of the default URLs?**

**Comment by rarkins** [comment]
npm is a little unique in Renovate in terms of auth and registries due to npm's `.npmrc` approach.

1. How are you running Renovate? e.g. installing the package from npmjs into a Node.js server, or running the official Renovate Docker image, or..?
2. Do you have any `.npmrc` files either (a) on the same system that Renovate runs on, or (b) committed to the repository itself?

**Comment by rarkins** [comment]
The above remarks are all still valid, however I there's something I missed previously which could explain things:

> `"//private_npm_registyr_url_with_token",`

Renovate's `registryUrls` is a generic feature and expects `https://` URLs.

Assuming you *don't* have any `.npmrc` files interfering, your config should look like this:

```json
{
  "packageRules": [{
    "datasources": ["npm"],
    "registryUrls": ["https://company.artifact.registry"]
  }],
  "hostRules": [{
    "baseUrl": "https://company.artifact.registry",
    "token": "the_registry_token"
  }]
}
```

The package rule above is essentially the same as the `npm: {}` syntax but I recommend using package rules where possible. Then the host rule is what tells our request handler to add a token to outgoing requests.

I'm also looking into whether we can do stronger validation of `registryUrls` to avoid this problem or at least warn about it.

**Comment by jsrice7391** [comment]
@rarkins thanks for the feedback, I will try it out. The intent is to run it on a scheduled job within CircleCI and then have repositories within the company opt-in by adding their repo to the repositories array. Right now, I have `renovate` within the `package.json`, run `yarn`,  and then run the scan by `./node_mdules/renovate/dist/renovate.js` (limitations on the container when it comes to accessing global npm installs).

The `.npmrc` file is created by grabbing the CodeArtifact token before the scan:

```bash
export CODEARTIFACT_TOKEN=$(aws codeartifact get-authorization-token --region us-east-1 --domain my_domain --domain-owner 123456789 --query authorizationToken --output text)
export PYPI_URL="https://aws:$CODEARTIFACT_TOKEN@my_domain-123456789.d.codeartifact.us-east-1.amazonaws.com/pypi/my_repo/simple/"
export CODE_ARTIFACT_REPOSITORY_URL="https://my_domain-123456789.d.codeartifact.us-east-1.amazonaws.com/npm/my_repo/"
export NPM_REGISTRY="//my_domain-123456789.d.codeartifact.us-east-1.amazonaws.com/npm/my_repo"
npm config set registry="$CODE_ARTIFACT_REPOSITORY_URL"
npm config set $NPM_REGISTRY/:_authToken=$CODEARTIFACT_TOKEN
npm config set $NPM_REGISTRY/:always-auth=true
npm -d ping --registry $CODE_ARTIFACT_REPOSITORY_URL
```

**Reply by rarkins** [reply]
A `.npmrc` file is an alternative to packageRules + hostRules so you should use either only `.npmrc` or only the two rules. If it's most convenient to use the `.npmrc` then go with that option. Is it being written to `$HOME/.npmrc`?

**Reply by jsrice7391** [reply]
I like the idea of keeping that script and then copying the `~/.npmrc` to the project directory before scanning. That works. Still getting the same error. According to this doc, if I place a `.npmrc` file within the project directory and I do not need to add anything authentication related to the project directory?

**Reply by rarkins** [reply]
If there are `.npmrc` files in either the repository or within the file system then we try to replicate npm's logic for these, because when it comes time to ask npm to update the lock file then it will use its own registry and auth resolution anyway.

The cleanest way to do it is to have zero `.npmrc` files in either the destination repositories or in your machine, and instead define an `npmrc` field within your bot config (e.g. `config.js`).

The first step will be Renovate's own lookup, which it does using http directly and not using `npm` itself at all. However if a `.npmrc` is present then it calls an npm lib to check for that and use it.

So best thing is to take a step back and assess:
- Do you have any `.npmrc` in the repository?
- Do you have any `.npmrc` in the file system? If so then where?
- Have you added any `npmrc` or `npmToken` fields to your Renovate config?
- Have you added any `packageRules` or `hostRules` to your Renovate config?

**Comment by dkrowl** [comment]
@rarkins I'm continuing this topic as it's not only npm.

I'm facing issues with gradle and the same problem. I have a Nexus-repository group with proxy and hosted repositories. In my gradle project I have set:
* pluginManagement repositories to private repository.
* build.gradle buildscript repositories to private repository.
* build.gradle repositories to private repository.

and

`  "packageRules": [
    {
      "matchDatasources": ["gradle"],
      "registryUrls": ["https://domain/repository/reponame/"]
    }
  ]
`

I'm running renovate with the latest renovate docker-image. And in the logs I can see that it is trying to search plugins and libraries from https://repo.maven.apache.org/maven2/ and https://plugins.gradle.org/m2/. Why? Nothing in the project settings should direct to those repositories and even with the packageRules it's still doing it. Eventually it works but it's super slow as renovate is waiting for those addresses to answer.
